File: languages/js/src/client/encryption.ts
import { BloockBridge } from "../bridge/bridge";
import { ConfigData } from "../bridge/proto/config";
import { KeyPair, RsaKeyPair } from "../entity/key";
import { Record } from "../entity/record/record";
import { NewConfigData } from "../config/config";
import {
  EncryptionAlgRequest,
  EncryptRequest,
  DecryptRequest
} from "../bridge/proto/encryption";
import { EncryptionAlg, Encrypter } from "../entity/encryption";
import { GenerateLocalKeyRequest } from "../bridge/proto/keys";
import { KeyType } from "../bridge/proto/keys_entities";

/**
 * Represents a client for interacting with the [Bloock Encryption service](https://dashboard.bloock.com/login).
 */
export class EncryptionClient {
  private bridge: BloockBridge;
  private configData: ConfigData;

  /**
   * Creates a new instance of the EncryptionClient with default configuration.
   * @param configData 
   */
  constructor(configData?: ConfigData) {
    this.bridge = new BloockBridge();
    this.configData = NewConfigData(configData);
  }

  /**
   * GenerateRsaKeyPair generates an RSA key pair for encryption.
   * @returns {Promise<KeyPair>} An object containing both the public and the private key
   * @deprecated Will be deleted in future versions. Use KeyClient.newLocalKey function instead.
   */
  public async generateRsaKeyPair(): Promise<KeyPair> {
    let request = GenerateLocalKeyRequest.fromPartial({
      configData: this.configData,
      keyType: KeyType.Rsa2048
    });

    return this.bridge
      .getKey()
      .GenerateLocalKey(request)
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return RsaKeyPair.fromProto(res);
      });
  }

  /**
   * Encrypts a Bloock record using the specified encrypter.
   * @param record 
   * @param encrypter 
   * @returns 
   */
  async encrypt(record: Record, encrypter: Encrypter): Promise<Record> {
    return this.bridge
      .getEncryption()
      .Encrypt(
        EncryptRequest.fromPartial({
          configData: this.configData,
          record: record.toProto(),
          encrypter: encrypter.toProto()
        })
      )
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return Record.fromProto(res.record!, this.configData);
      });
  }

  /**
   * Decrypts a Bloock record using the specified decrypter.
   * @param record 
   * @param decrypter 
   * @returns 
   */
  async decrypt(record: Record, decrypter: Encrypter): Promise<Record> {
    return this.bridge
      .getEncryption()
      .Decrypt(
        DecryptRequest.fromPartial({
          configData: this.configData,
          record: record.toProto(),
          decrypter: decrypter.toProto()
        })
      )
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return Record.fromProto(res.record!, this.configData);
      });
  }

  /**
   * Gets the encryption algorithm used for a Bloock record.
   * @param record 
   * @returns 
   */
  async getEncryptionAlg(record: Record): Promise<EncryptionAlg> {
    return this.bridge
      .getEncryption()
      .GetEncryptionAlg(
        EncryptionAlgRequest.fromPartial({
          configData: this.configData,
          record: record.toProto()
        })
      )
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return EncryptionAlg.fromProto(res.alg);
      });
  }
}

File: languages/go/client/encryption.go
package client

import (
	"context"
	"errors"

	"github.com/bloock/bloock-sdk-go/v2/entity/encryption"
	"github.com/bloock/bloock-sdk-go/v2/entity/key"
	"github.com/bloock/bloock-sdk-go/v2/entity/record"
	"github.com/bloock/bloock-sdk-go/v2/internal/bridge"
	"github.com/bloock/bloock-sdk-go/v2/internal/bridge/proto"
	"github.com/bloock/bloock-sdk-go/v2/internal/config"
)

// EncryptionClient represents a client for interacting with the [Bloock Encryption service].
//
// [Bloock Encryption service]: https://bloock.com
type EncryptionClient struct {
	bridgeClient bridge.BloockBridge
	configData   *proto.ConfigData
}

// NewEncryptionClient creates a new instance of the EncryptionClient with default configuration.
func NewEncryptionClient() EncryptionClient {
	return EncryptionClient{
		bridgeClient: bridge.NewBloockBridge(),
		configData:   config.NewConfigDataDefault(),
	}
}

// NewEncryptionClientWithConfig creates a new instance of the EncryptionClient with the provided configuration.
func NewEncryptionClientWithConfig(configData *proto.ConfigData) EncryptionClient {
	return EncryptionClient{
		bridgeClient: bridge.NewBloockBridge(),
		configData:   configData,
	}
}

// GenerateRsaKeyPair generates an RSA key pair for encryption.
// Deprecated: Will be deleted in future versions. Use KeyClient.newLocalKey function instead.
func (c *EncryptionClient) GenerateRsaKeyPair() (key.KeyPair, error) {
	res, err := c.bridgeClient.Key().GenerateLocalKey(context.Background(), &proto.GenerateLocalKeyRequest{
		ConfigData: c.configData,
		KeyType:    key.KeyTypeToProto[key.Rsa2048],
	})

	if err != nil {
		return key.KeyPair{}, err
	}

	if res.Error != nil {
		return key.KeyPair{}, errors.New(res.Error.Message)
	}

	return key.NewRsaKeyPairFromProto(res), nil
}

// Encrypt encrypts a Bloock record using the specified encrypter.
func (c *EncryptionClient) Encrypt(r record.Record, encrypter encryption.Encrypter) (record.Record, error) {
	res, err := c.bridgeClient.Encryption().Encrypt(context.Background(), &proto.EncryptRequest{
		ConfigData: c.configData,
		Record:     r.ToProto(),
		Encrypter:  encrypter.ToProto(),
	})

	if err != nil {
		return record.Record{}, err
	}

	if res.Error != nil {
		return record.Record{}, errors.New(res.Error.Message)
	}

	return record.NewRecordFromProto(res.Record, c.configData), nil
}

// Decrypt decrypts a Bloock record using the specified decrypter.
func (c *EncryptionClient) Decrypt(r record.Record, decrypter encryption.Encrypter) (record.Record, error) {
	res, err := c.bridgeClient.Encryption().Decrypt(context.Background(), &proto.DecryptRequest{
		ConfigData: c.configData,
		Record:     r.ToProto(),
		Decrypter:  decrypter.ToProto(),
	})

	if err != nil {
		return record.Record{}, err
	}

	if res.Error != nil {
		return record.Record{}, errors.New(res.Error.Message)
	}

	return record.NewRecordFromProto(res.Record, c.configData), nil
}

// GetEncryptionAlg retrieves the encryption algorithm used for a Bloock record.
func (c *EncryptionClient) GetEncryptionAlg(r record.Record) (encryption.EncryptionAlg, error) {
	bridgeClient := bridge.NewBloockBridge()
	res, err := bridgeClient.Encryption().GetEncryptionAlg(context.Background(), &proto.EncryptionAlgRequest{ConfigData: c.configData, Record: r.ToProto()})

	if err != nil {
		return encryption.UNRECOGNIZED_ENCRYPTION_ALG, err
	}

	if res.Error != nil {
		return encryption.UNRECOGNIZED_ENCRYPTION_ALG, errors.New(res.Error.Message)
	}

	return encryption.EncryptionAlgFromProto[res.Alg], nil
}
